pkg/controllers/termination/terminator/terminator.go

```diff
@@ -59,60 +61,61 @@ func (t *Terminator) Cordon(ctx context.Context, node *v1.Node) error {
 		if err := t.kubeClient.Patch(ctx, node, client.MergeFrom(stored)); err != nil {
 			return err
 		}
-		logging.FromContext(ctx).Infof("cordoned node")
+		logging.FromContext(ctx).Infof("cordoning")
 	}
 	return nil
 }
 
 // Drain evicts pods from the node and returns true when all pods are evicted
 // https://kubernetes.io/docs/concepts/architecture/nodes/#graceful-node-shutdown
 func (t *Terminator) Drain(ctx context.Context, node *v1.Node) error {
-	// Get evictable pods
-	pods, err := t.getPods(ctx, node)
-	if err != nil {
-		return fmt.Errorf("listing pods for node, %w", err)
+	// Grace period is up, we can no longer drain
+	if t.clock.Now().After(node.DeletionTimestamp.Time) {
+		return nil
 	}
-	var podsToEvict []*v1.Pod
-	// Skip node due to pods that are not able to be evicted
-	for _, p := range pods {
-		// Ignore if unschedulable is tolerated, since they will reschedule
-		if podutil.ToleratesUnschedulableTaint(p) {
-			continue
-		}
-		// Ignore static mirror pods
-		if podutil.IsOwnedByNode(p) {
-			continue
-		}
-		podsToEvict = append(podsToEvict, p)
-	}
-	// Enqueue for eviction
-	t.evict(podsToEvict)
-
-	if len(podsToEvict) > 0 {
-		return NewNodeDrainError(fmt.Errorf("%d pods are waiting to be evicted", len(podsToEvict)))
-	}
-	return nil
-}
 
-// getPods returns a list of evictable pods for the node
-func (t *Terminator) getPods(ctx context.Context, node *v1.Node) ([]*v1.Pod, error) {
 	podList := &v1.PodList{}
 	if err := t.kubeClient.List(ctx, podList, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {
-		return nil, fmt.Errorf("listing pods on node, %w", err)
+		return fmt.Errorf("listing pods on node, %w", err)
 	}
-	var pods []*v1.Pod
-	for _, p := range podList.Items {
-		// Ignore if the pod is complete and doesn't need to be evicted
-		if podutil.IsTerminal(lo.ToPtr(p)) {
-			continue
+	pods := lo.Map(podList.Items, func(pod v1.Pod, _ int) *v1.Pod { return lo.ToPtr(pod) })
+	evictable := lo.Reject(pods, func(p *v1.Pod, _ int) bool {
+		return podutil.IsTerminal(p) ||
+			(podutil.IsTerminating(p) && t.clock.Now().After(p.DeletionTimestamp.Time.Add(1*time.Minute))) ||
```

Comment:
Why do we need to add a minute?
```suggestion
			(podutil.IsTerminating(p) && t.clock.Now().After(p.DeletionTimestamp.Time))) ||
```

Reply:
This is "stuck terminating". I just kept the minute that was already implemented. I assume it's to catch race conditions, but not totally sure.